File: android/src/main/java/com/jvanila/flutter/shopify/usecases/EditCustomerInfoUseCase.java
package com.jvanila.flutter.shopify.usecases;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.jvanila.flutter.plugin.arch.PluginContext;
import com.shopapp.gateway.ApiCallback;
import com.shopapp.gateway.entity.Customer;
import com.shopapp.gateway.entity.Error;
import com.shopapp.shopify.api.ShopifyApi;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class EditCustomerInfoUseCase extends ShopifyCallUseCase {

    private static final String ARG_FIRST_NAME = "firstName";
    private static final String ARG_LAST_NAME = "lastName";
    private static final String ARG_PHONE = "phone";
    private static final String ARG_EMAIL = "email";


    public EditCustomerInfoUseCase(PluginContext<ShopifyApi> pluginContext) {
        super(pluginContext);
    }

    @Override
    protected void call(MethodCall input, final MethodChannel.Result result) {
        String firstName = input.argument(ARG_FIRST_NAME);
        String lastName = input.argument(ARG_LAST_NAME);
        String phone = input.argument(ARG_PHONE);
        String email = input.argument(ARG_EMAIL);

        mPluginContext.api.instance.editCustomerInfo(firstName, lastName, phone, email,
                new ApiCallback<Customer>() {
            @Override
            public void onResult(Customer customer) {
                if (customer != null) {

                    String customerJsonString = new Gson().toJson(customer,
                            new TypeToken<Customer>() {
                            }.getType());

                    result.success(customerJsonString);
                }
            }

            @Override
            public void onFailure(Error error) {
                System.out.println("onFailure -- " + error);
                result.error("EditCustomerInfoUseCase", error.getMessage(), error);
            }
        });
    }
}

File: ios/Classes/UseCases/EditCustomerInfoUseCase.swift
//
//  EditCustomerInfoUseCase.swift
//  shopify
//
//  Created by Joseph jose on 10/09/19.
//

import UIKit
import ShopApp_Shopify
import ShopApp_Gateway
import Flutter

class EditCustomerInfoUseCase: UseCase {

    static let ARG_FIRST_NAME = "firstName"
    static let ARG_LAST_NAME = "lastName"
    static let ARG_PHONE = "phone"
    static let ARG_EMAIL = "email"
    
    override public init(_ context: PluginContext) {
        super.init(context)
    }
    
    override public func trigger(with methodCall: FlutterMethodCall, result: @escaping (Any?) -> Void) {
        
        if let args = methodCall.arguments as? [String:String] {

            let firstName = args[EditCustomerInfoUseCase.ARG_FIRST_NAME]
            let lastName = args[EditCustomerInfoUseCase.ARG_LAST_NAME]
            let email = args[EditCustomerInfoUseCase.ARG_EMAIL]
            let phone = args[EditCustomerInfoUseCase.ARG_PHONE]
            
            (mContext.api.instance as! ShopifyAPI).updateCustomer(with: email!, firstName: firstName, lastName: lastName, phone: phone) { (customer, error) in
                
                if error != nil {
                    
                    self.createError(withCase: "EditCustomerInfoUseCase", message: error!.errorMessage, error: error!, on: result)
                }else {
                    
                    let customerJSONObj = customer!.toDictionary()
                    do {
                        
                        let customerJSON = try JSONSerialization.data(withJSONObject: customerJSONObj, options: .prettyPrinted)
                        let customerJsonString = String(data: customerJSON, encoding: .utf8)
                        result(customerJsonString)
                    }
                    catch {
                        
                        self.createError(withCase: "EditCustomerInfoUseCase", message: error.localizedDescription, error: error, on: result)
                    }
                }
            }
        }
    }
}
